File: sortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        
        ListNode *temp = head;
        vector<int> nums;
        
        while(temp != NULL)
        {
            nums.push_back(temp -> val);
            temp = temp -> next;
        }
        
        temp = head;
        
        sort(nums.begin(), nums.end());
        
        for(auto num: nums)
        {
            temp -> val = num;
            temp = temp -> next;
        }
        
        return head;
    }
};

File: progress_sheet/sortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if not head or not head.next:
            return head
        
        # Divide
        dummy = ListNode(0, head)
        slow = dummy
        fast = dummy

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        mid = slow.next
        slow.next = None
        
        left = self.sortList(head)
        right = self.sortList(mid)

        # Merge
        dummy = ListNode()
        curr = dummy

        while left and right:
            if left.val <= right.val:
                curr.next = left
                left = left.next
            
            else:
                curr.next = right
                right = right.next
            
            curr = curr.next
        
        if not left:
            curr.next = right
        
        else:
            curr.next = left
        
        return dummy.next
